amqp/Go/amqpProducer.go

```diff
@@ -1,218 +1,210 @@
+/*                                                                                      */
+/*  Parameters:                                                                         */
+/*  -t ==> TopicName                                                                    */
+/*  -q ==> QueueName                                                                    */
+/****************************************************************************************/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/Azure/go-amqp"
+)
+
+type Env struct {
+	Username string `json:"APP_USER"`
+	Password string `json:"APP_PASSWORD"`
+	Host     string `json:"HOST"`
+	Port     string `json:"PORT"`
+}
+
+type MQEndpoints struct {
+	Points []Env `json:"MQ_ENDPOINTS"`
+}
+
+var (
+	queueName   string
+	topicName   string
+	destination string
+	isQueue     bool
+)
+var EnvSettings Env
+var MQ_ENDPOINTS MQEndpoints
+
+// Parsing the command line arguments
+func parseArgs() {
+	flag.StringVar(&queueName, "q", "", "Queue")
+	flag.StringVar(&topicName, "t", "", "Topic")
+	flag.Parse()
+
+	// One, and only one, of the parameters must be provided
+	if (queueName == "" && topicName == "") ||
+		(queueName != "" && topicName != "") {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	if queueName != "" {
+		isQueue = true
+		destination = queueName
+	} else {
+		destination = topicName
+	}
+}
+
+func main() {
+	parseArgs()
+	parseEnv()
+
+	fmt.Println("Application is Starting...")
+
+	//Defining the AMQP URI
+	uri := "amqp://" + EnvSettings.Username + ":" + EnvSettings.Password + "@" + EnvSettings.Host + ":" + EnvSettings.Port
+
+	// Set up a new connection
+	conn, err := connect(uri)
+	if err != nil {
+		log.Fatal("Error in creating connection: ", err)
+	}
+
+	// Create a session
+	session, err := createSession(conn)
+	if err != nil {
+		log.Fatal("Error in creating a session: ", err)
+	}
+
+	//When the destination is a queue, capability must be set to "queue". If not, topic is selected by default.
+	SenderOptions := &amqp.SenderOptions{
+		TargetCapabilities: []string{"topic"},
+	}
+	if isQueue {
+		SenderOptions = &amqp.SenderOptions{
+			TargetCapabilities: []string{"queue"},
+		}
+	}
+
+	// Create a sender
+	sender, err := createSender(session, SenderOptions)
+	if err != nil {
+		log.Fatal("Error in creating a sender: ", err)
+	}
+
+	//Message.Properties.To must contain the destination address (URI)
+	addr := uri
+
+	// Message to be sent
+	msg := &amqp.Message{
+		Data: [][]byte{[]byte("Hello, World!")},
+		Properties: &amqp.MessageProperties{
+			To: &addr,
+		},
+	}
+
+	// Send the messages
+	for i := 0; i < 10; i++ {
+		err = sendMessage(sender, msg)
+		if err != nil {
+			log.Fatal("Error in sending messages: ", err)
+		}
+	}
+
+	log.Println("Messages sent successfully!")
+	fmt.Println("End of Sample amqpProducer.go")
+}
+
+func connect(uri string) (*amqp.Conn, error) {
```

Comment:
createSession and connect are common, so a chance to refactor the code to remove code duplication.

Reply:
There's actually even more that's common - parseEnv(), parseArgs(). I did consider a more major rework - perhaps putting the whole thing into a single program and switching based on a flag - but didn't really want to do a redesign, just fix the most egregious issues.